File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.cpp
class Solution {
public:

    int countSubarraysSumGreaterThanK(vector<int>& a, int k)
    {
        int n=a.size();
        int sum=0,count=0,i=0;
        for(int j=0;j<n;j++)
        {
            sum+=a[j];
            while(i<=j && sum>=k)   // a = [0,0,0,0,0]
            {
                count+=n-j;
                sum-=a[i];
                i++;
            }
        }
        return count;
    }

    int numSubarraysWithSum(vector<int>& a, int k) 
    {
        return countSubarraysSumGreaterThanK(a,k)-countSubarraysSumGreaterThanK(a,k+1);
    }
};

File: 966-binary-subarrays-with-sum/binary-subarrays-with-sum.java
class Solution {
    private int numSubarraysWithSumAtLeastK(int[] nums, int k) {
        int count = 0, i = 0, sum = 0;
        for (int j = 0; j < nums.length; j++) {
            sum += nums[j];
            while(sum >= k && i <= j) {
                count += nums.length - j;
                sum -= nums[i];
                i++;
            }
        }
        return count;
    }

    public int numSubarraysWithSum(int[] nums, int goal) 
    {
        return numSubarraysWithSumAtLeastK(nums, goal) - numSubarraysWithSumAtLeastK(nums, goal + 1);
    }
}